CMakeLists.txt

```diff
@@ -26,6 +20,16 @@ OPTION(BUILD_STATIC_LIB "Build static library" OFF)
 OPTION(__BUILD_FOR_R "Set to ON if building lib_lightgbm for use with the R package" OFF)
 OPTION(__INTEGRATE_OPENCL "Set to ON if building LightGBM with the OpenCL ICD Loader and its dependencies included" OFF)
 
+if(__BUILD_FOR_R)
+  set(CMAKE_TRY_COMPILE_TARGET_TYPE "STATIC_LIBRARY")
```

Comment:
It seems that `CMAKE_TRY_COMPILE_TARGET_TYPE` is available only in CMake >= 3.6.
https://cmake.org/cmake/help/v3.19/variable/CMAKE_TRY_COMPILE_TARGET_TYPE.html
We should add 
```
if(__BUILD_FOR_R)
  cmake_minimum_required(VERSION 3.6)
elseif(__INTEGRATE_OPENCL)
...
```
here, I believe.
https://github.com/microsoft/LightGBM/blob/71a1a4f6b2ef2bf5074fcfd9bbce0832366f1861/CMakeLists.txt#L1-L9

Reply:
Oh ok sure! I just added it in https://github.com/microsoft/LightGBM/pull/3541/commits/4e21bbd0b52380b473c07853093e31fdaaf372ab

I also moved all those checks under the `OPTION()` calls. It seemed strange to me that we're doing things like `IF(__INTEGRATE_OPENCL)` (for example) before `OPTION(__INTEGRATE_OPENCL)`. I think that's a mistake. If you look at https://cmake.org/cmake/help/latest/command/if.html#condition-syntax, it says

> `if(<variable|string>)`
> True if given a variable that is defined to a value that is not a false constant. False otherwise. (Note macro arguments are not variables.)

I think that `IF(THING_THAT_HAS_NOT_BEEN_DEFINED)` will always evaluate to false.

Reply:
> I think that ``IF(THING_THAT_HAS_NOT_BEEN_DEFINED)`` will always evaluate to false.

I'm not sure but it seems that we define "``THING_THAT_HAS_NOT_BEEN_DEFINED``" during passing `-DTHING_THAT_HAS_NOT_BEEN_DEFINED` CMake flag. For example, I remember everything was working OK even without `OPTION` at all: https://github.com/microsoft/LightGBM/pull/3144#discussion_r487467750.

https://cmake.org/cmake/help/v3.0/manual/cmake.1.html#options

Reply:
Right, makes sense. I think we just got lucky that all options that were used in code before the `OPTION()` calls have default value `OFF`, which behaves the same as it not being defined at all.

Anyway, I'm glad that they're all moved to the top now.